src/combinatorics/derangements/finite.lean

```diff
@@ -0,0 +1,117 @@
+
+  - `num_derangements n`: The number of derangements on an n-element set, defined in a computation-
+    friendly way.
+
+  - `card_derangements_eq_num_derangements`: Proof that `num_derangements` really does compute the
+    number of derangements.
+
+  - `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
+    factorials.
+-/
+
+open derangements equiv fintype
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (derangements α) := λ _, fintype.decidable_forall_fintype
+
+instance : fintype (derangements α) := by delta_instance derangements
+
+lemma card_derangements_invariant {α β : Type*} [fintype α] [decidable_eq α]
+  [fintype β] [decidable_eq β] (h : card α = card β) :
+  card (derangements α) = card (derangements β) :=
+fintype.card_congr (equiv.derangements_congr $ equiv_of_card_eq h)
+
+lemma card_derangements_fin_add_two (n : ℕ) :
+  card (derangements (fin (n+2))) = (n+1) * card (derangements (fin n)) +
+  (n+1) * card (derangements (fin (n+1))) :=
+begin
+  -- get some basic results about the size of fin (n+1) plus or minus an element
+  have h1 : ∀ a : fin (n+1), card ({a}ᶜ : set (fin (n+1))) = card (fin n),
+  { intro a,
+    simp only [fintype.card_fin, finset.card_fin, fintype.card_of_finset, finset.filter_ne' _ a,
+      set.mem_compl_singleton_iff, finset.card_erase_of_mem (finset.mem_univ a), nat.pred_succ] },
+  have h2 : card (fin (n+2)) = card (option (fin (n+1))),
+  { simp only [card_fin, card_option] },
+  -- rewrite the LHS and substitute in our fintype-level equivalence
+  simp only [card_derangements_invariant h2,
+    card_congr (@derangements_recursion_equiv (fin (n+1)) _),
+  -- push the cardinality through the Σ and ⊕ so that we can use `card_n`
+    card_sigma, card_sum, card_derangements_invariant (h1 _), finset.sum_const, nsmul_eq_mul,
+    finset.card_fin, mul_add, nat.cast_id],
+end
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements : ℕ → ℕ
+| 0 := 1
+| 1 := 0
+| (n + 2) := (n + 1) * (num_derangements n + num_derangements (n+1))
+
+@[simp] lemma num_derangements_zero : num_derangements 0 = 1 := rfl
+
+@[simp] lemma num_derangements_one : num_derangements 1 = 0 := rfl
+
+lemma num_derangements_add_two (n : ℕ) :
+  num_derangements (n+2) = (n+1) * (num_derangements n + num_derangements (n+1)) := rfl
+
+lemma num_derangements_succ (n : ℕ) :
+  (num_derangements (n+1) : ℤ) = (n + 1) * (num_derangements n : ℤ) - (-1)^n :=
+begin
+  induction n with n hn,
+  { refl },
+  { simp only [num_derangements_add_two, hn, pow_succ,
+      int.coe_nat_mul, int.coe_nat_add, int.coe_nat_succ],
+    ring }
+end
+
+lemma card_derangements_fin_eq_num_derangements {n : ℕ} :
+  card (derangements (fin n)) = num_derangements n :=
+begin
+  apply nat.strong_induction_on n, clear n, -- to avoid confusion with the n in the hypothesis
+  rintros (_|_|n) hyp, { refl }, { refl },  -- knock out cases 0 and 1
```

Comment:
I think we'd generally use
```suggestion
  induction n using nat.strong_induction_on with n hyp,
  obtain (_|_|n) := n, { refl }, { refl },  -- knock out cases 0 and 1
```

Reply:
aha, i figured there was some way of using nat.strong_induction_on that was better than what i was doing. i'll have to write that down somewhere.